File: axpy/axpy.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU Mathematics
! Math 6370
! 22 February 2009
!=================================================================


program axpy
  !-----------------------------------------------------------------
  ! Description: 
  !    Performs some simple vector linear combinations.
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer, parameter :: MAX_INTEGER_DIGITS = 10
  integer :: n, i
  character(MAX_INTEGER_DIGITS) :: n_string
  double precision :: a, stime, ftime
  double precision, allocatable :: x(:), y(:), z(:)

  !======= Internals ============
  
  ! get n from the command line
  call getarg(1, n_string)

  ! ensure that an argument was passed in
  if ( n_string == '' ) then
     stop 'Error: function requires one argument (vector length)'
  endif
  
  ! convert n_string to integer
  read (n_string, *) n

  ! allocate arrays
  allocate(x(n), y(n), z(n))

  ! start timer
  call get_time(stime)

  ! initialize a, x and y
  a = -3.d0
  do i=1,n
     x(i) = exp(2.d0*i/n)
     y(i) = 1.d0*(n-1)/n
  end do

  ! perform linear combination
  do i=1,n
     z(i) = a*x(i) + y(i)
  end do
  
  do i=1,n
     x(i) = y(i)/a - z(i)
  end do
  
  do i=1,n
     y(i) = x(i)*y(i)/n
  end do

  ! output maximum value in z
  write(*,*) '  max(z) =',maxval(z)

  ! stop timer
  call get_time(ftime)

  ! output total time
  write(*,*) ' runtime =',ftime-stime
  
  ! deallocate arrays
  deallocate(x, y, z)


end program axpy
!=================================================================

File: axpy/axpy.c
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370 / 6370 */


/* Inclusions */
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "get_time.h"


/* Example routine to perform some simple vector linear combinations. */
int main(int argc, char* argv[]) {

  /* set the vector length */
  int n, i;
  double *x, *y, *z, a, zmax, runtime;
  double stime, ftime;
  
   // ensure that an argument was passed in
  if (argc < 2) {
    printf("Error: function requires one argument (vector length)\n");
    return 1;
  }

  // set n as the input argument, and ensure it's positive
  n = atoi(argv[1]);
  if (n < 1) {
    printf("Error: vector length %i must be greater than 0\n", n);
    return 1;
  }

  /* allocate the vectors */
  x = malloc(n*sizeof(double));
  y = malloc(n*sizeof(double));
  z = malloc(n*sizeof(double));

  /* start timer */
  stime = get_time();

  /* initialize a, x and y */
  a = -3.0;
  for (i=0; i<n; i++) {
    x[i] = exp(2.0*(i+1)/n);
    y[i] = 1.0*(n-1)/n;
  }

  /* perform linear combinations */
  for (i=0; i<n; i++)  z[i] = a*x[i] + y[i];
  for (i=0; i<n; i++)  x[i] = y[i]/a - z[i];
  for (i=0; i<n; i++)  y[i] = x[i]*y[i]/n;

  /* output maximum value in z */
  zmax = z[0];
  for (i=0; i<n; i++) zmax = (zmax > z[i]) ? zmax : z[i];
  printf("  max(z) = %.16e\n",zmax);

  /* stop timer */
  ftime = get_time();
  runtime = ftime - stime;

  /* output total time */
  printf(" runtime = %.16e\n",runtime);

  /* free vectors */
  free(x);
  free(y);
  free(z);

  return 0;
} /* end main */

